rfd/0179-api-versioning.md

```diff
@@ -0,0 +1,392 @@
+`GET /webapi/foos` , you can mark for deletion in v18, as v18 only needs to
+support n-1 versioned clients. It is the responsibility of the developer who is
+"removing the use" (i.e., making it obsolete with some other endpoint) to mark
+the TODO.
+
+
+### Adding/removing a field to a response of an existing endpoint
+Adding a new field to a response is OK as long as that field has to effect on
+the previously existing fields. Any field that was previously in a response
+should _stay populated_ in the new response, even if that creates duplicate
+data. An existing field _cannot_ have its type changed. A field should not be
+removed until the backward compatibility promise has been 'fulfilled'. If v17
+clients are only using a new field, you can mark the old field as "TODO DELETE"
+in v18. This means that v17 servers can support v16 clients (still using the old
+field) and v18 only needs to support v17 (using the new field).
+
+Bad
+```go
+type GetFoosResponse struct {
+    Foos []string
+    NextKey string
+    // ContainsDuplicates bool <--- previously existing field made redundant by DuplicateCount
+    DuplicateCount int
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     // somedev: "we don't need ContainsDuplcates anymore because now we can
+	     // just check if duplicate count > 0!"
+	     DuplicateCount: 1 
+	}, nil
+}
+```
+
+Good
+```go
+type GetFoosResponse struct {
+    Foos []string
+    NextKey string
+    // TODO (avatus) DELETE IN 18
+    ContainsDuplicates bool
+    DuplicateCount int
+}
+
+func (h *Handler) getFoos() (interface{}, error) {
+    foos := []string{"foo1", "foo2", "foo3", "foo3"}
+	return GetFoosResponse{
+	     Foos: foos,
+	     // somedev: "we must keep ContainsDuplicates populated to support older clients!"
+	     ContainsDuplicates: true
+	     DuplicateCount: 1 
+	}, nil
+}
+``` 
+If using ConnectRPC (a proposed solution later in this RFD), then we must _not_ remove fields from requests/responses, even if unused. Mark as deprecated and move on.
+
+If a response shape must change entirely, prefer creating a new endpoint/RPC.
+
+### Updating Request body fields
+A Request body can receive a new field as long as the omission of the new field
+would not cause a break (i.e., an old client sending the old request without the
+new field). If the handler _requires_ this new field in order to function, then
+a new endpoint must be created with a version increment.
+
+
+### Adding new features to the web client
+If a feature requires a new endpoint, it must properly handle the case of a 404.
+This means making the "expected" 404 error clear to the user that an endpoint is
+not found due to a version mismatch (and preferably which minimum version is
+needed) rather than an ambiguous "Not Found" error.
+
+### New required fields from an API response in the web client
```

Comment:
This is about the response shape. If the web client expects some _new_ field in the response, it needs to be calling a new version.

Reply:
I believe this would be handled with the `Updating request body fields` section. I will add your example and a disclaimer about new version should be a default!

Reply:
omg, i just now saw/understood the difference. `response` vs `request` 🤦‍♀️